### PR TITLE
[#53] scaledown should try a different pod if the current target is failing

### LINKED ISSUE #53: scaledown should try a different pod if the current target is failing
**Describe the bug**
Currently we always scaledown to pod 0 no matter whether the scaledown is successful or not.
It should in the fail case try a different pod (if available) until all are tried.

**To Reproduce**
1. Get X version of the source
2. '...'

**Platform (please complete the following information):**
 - OS: [e.g. Fedora]
 - Version [e.g. 32]
 - CEKit version [e.g. 3.5.0]
 - Container build tool [e.g. podman]

**Expected behavior**
Optional, use if not obvious. A clear and concise description of what you expected to happen.

**Screenshots**
Optional, add screenshots to help explain your problem.

**Additional context**
Optional. Add any other context about the problem here.

### PR DESCRIPTION
**Pull request by gaohoward**
- Now the drain.sh will try next pod if the first pod failed scaled down
- drain.sh logs are timestamped
- drain.sh logs will report time of scaledown and total message migrated.

Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:

- [ ] Pull Request title is properly formatted: `[CLOUD-XYA] Subject`
- [ ] Pull Request contains link to the JIRA issue
- [ ] Pull Request contains description of the issue
- [ ] Pull Request does not include fixes for issues other than the main ticket
- [ ] Attached commits represent units of work and are properly formatted
- [ ] You have read and agreed to the Developer Certificate of Origin (DCO) (see `CONTRIBUTING.md`)
- [ ] Every commit contains `Signed-off-by: Your Name <yourname@example.com>` - use `git commit -s`

### CODE REVIEW

File: modules/activemq-artemis-launch/added/drain.sh [hunk 1]
```diff
@@ -1,177 +1,244 @@
...
+)
 
-endpointsUrl="https://${KUBERNETES_SERVICE_HOST:-kubernetes.default.svc}:${KUBERNETES_SERVICE_PORT:-443}/api/v1/namespaces/${POD_NAMESPACE}/"
-endpointsAuth="Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)"
+  all_addresses=$(echo "$RET_VALUE" | python2 -c "$PYCMD")
+  log "get back all addresses $all_addresses"
+  arr=($all_addresses)
+
+  TOTAL_MESSAGES_ON_BROKER=0
+  for address in "${arr[@]}"
```
Comment (brusdev) [inline modules/activemq-artemis-launch/added/drain.sh]: Installing `jq` or `yq` the python scripts could be replaced by simpler commands, i.e.
```suggestion
  all_addresses=$(echo "$RET_VALUE" | jq -r '.value[]')
  log "get back all addresses $all_addresses"

  TOTAL_MESSAGES_ON_BROKER=0
  for address in "${all_addresses}"
```

File: modules/activemq-artemis-launch/added/drain.sh [hunk 2]
```diff
@@ -1,177 +1,244 @@
...
+  do
+    log "checking on address ${address}"
+    M_COUNT=$(curl -s -G -k http://${AMQ_USER}:${AMQ_PASSWORD}@${BROKER_HOST}:8161/console/jolokia/read/org.apache.activemq.artemis:broker=%22${AMQ_NAME}%22,address=%22${address}%22,component=addresses/MessageCount)
+    value=$(echo $M_COUNT | python2 -c "import sys, json; print json.load(sys.stdin)['value']")
```
Comment (brusdev) [inline modules/activemq-artemis-launch/added/drain.sh]: Installing jq or yq the python scripts could be replaced by simpler commands, i.e.
```suggestion
    value=$(echo $M_COUNT | jq -r '.value')
```

File: modules/activemq-artemis-launch/added/drain.sh [hunk 3]
```diff
@@ -1,177 +1,244 @@
...
-  echo "[drain.sh] Can't find target host name"
-  exit 1
-fi
+    SCALE_TO_BROKER="${hostName}"
+    log "scale down target is: $SCALE_TO_BROKER"
 
-source /opt/amq/bin/launch.sh nostart
+    # Add connector to the pod to scale down to
+    log "removing any existing scaledownconnector"
+    sed -i "s/<connector name=\"scaledownconnector\">.*//" ${instanceDir}/etc/broker.xml
```
Comment (brusdev) [inline modules/activemq-artemis-launch/added/drain.sh]: another alternative is using the delete command, i.e.
```suggestion
    sed -i '/<connector name="scaledownconnector">.*/d' ${instanceDir}/etc/broker.xml
```

File: modules/activemq-artemis-launch/added/drain.sh [hunk 4]
```diff
@@ -1,177 +1,244 @@
...
+    log "initiating scaledown. There are $total_before_scaledown messages to be migrated"
+    mm_start=$(date +%s)
 
-echo "[drain.sh] curl return code ${HTTP_CODE}"
+    RET_CODE=`curl -s -G -k http://${AMQ_USER}:${AMQ_PASSWORD}@${BROKER_HOST}:8161/console/jolokia/exec/org.apache.activemq.artemis:broker=%22${AMQ_NAME}%22/scaleDown/scaledownconnector`
 
-if [ "${HTTP_CODE}" != "200" ]; then
-  echo "[drain.sh] scaleDown is not successful, response: $RET_CODE"
-  echo "[drain.sh] sleeping for 30 seconds to allow inspection before it restarts"
-  sleep 30
```
Comment (brusdev) [inline modules/activemq-artemis-launch/added/drain.sh:134]: the previous script sleeps for 30 seconds after a failure to allow inspection, does the new script allows inspection?

File: modules/activemq-artemis-launch/added/drain.sh [hunk 5]
```diff
@@ -1,177 +1,244 @@
...
-endpointsAuth="Authorization: Bearer $(cat /var/run/secrets/kubernetes.io/serviceaccount/token)"
+  all_addresses=$(echo "$RET_VALUE" | python2 -c "$PYCMD")
+  log "get back all addresses $all_addresses"
+  arr=($all_addresses)
+
+  TOTAL_MESSAGES_ON_BROKER=0
+  for address in "${arr[@]}"
+  do
+    log "checking on address ${address}"
+    M_COUNT=$(curl -s -G -k http://${AMQ_USER}:${AMQ_PASSWORD}@${BROKER_HOST}:8161/console/jolokia/read/org.apache.activemq.artemis:broker=%22${AMQ_NAME}%22,address=%22${address}%22,component=addresses/MessageCount)
```
Comment (brusdev) [inline modules/activemq-artemis-launch/added/drain.sh]: this script counts all messages while the ActiveMQ Artemis ScaleDownHandler skips temporary queues, could this diff cause any issues?
https://github.com/apache/activemq-artemis/blob/2.31.2/artemis-server/src/main/java/org/apache/activemq/artemis/core/server/impl/ScaleDownHandler.java#L129

File: modules/activemq-artemis-launch/added/drain.sh (hunk 1 above)
Comment (gaohoward) [reply modules/activemq-artemis-launch/added/drain.sh]: ok, will use jq. Thanks!

File: modules/activemq-artemis-launch/added/drain.sh (hunk 2 above)
Comment (gaohoward) [reply modules/activemq-artemis-launch/added/drain.sh]: will do.

File: modules/activemq-artemis-launch/added/drain.sh (hunk 3 above)
Comment (gaohoward) [reply modules/activemq-artemis-launch/added/drain.sh]: ok, will use d command.

File: modules/activemq-artemis-launch/added/drain.sh (hunk 4 above)
Comment (gaohoward) [reply modules/activemq-artemis-launch/added/drain.sh:134]: The sleep is moved to when all the pods have tried and failed. here https://github.com/artemiscloud/activemq-artemis-broker-kubernetes-image/pull/81/files#diff-16b237f518d394fb048aa263553f35a607e7d6ae8ecf9bb9b2cb6a73a78c7a7fR97

File: modules/activemq-artemis-launch/added/drain.sh (hunk 5 above)
Comment (gaohoward) [reply modules/activemq-artemis-launch/added/drain.sh]: We can exclude temporary queues here to make the count precise. (by checking the Temporary attr). I'll update the code. Thanks!

Comment (gaohoward) [reply modules/activemq-artemis-launch/added/drain.sh]: @brusdev I've updated the code to exclude temporary queues.

**Review by brusdev — APPROVED** [review]
It LGTM, my only perplexity is on errors handling, i.e. the function `get_address_message_count` just logs the error and returns when a jolokia read fails so the caller ignores those errors.

**Comment by gaohoward** [comment]
> It LGTM, my only perplexity is on errors handling, i.e. the function `get_address_message_count` just logs the error and returns when a jolokia read fails so the caller ignores those errors.

Yes that's something need to be considered. I think we should be always on the safe side regarding scaledown. I'll add a check for that after scaledown returns successful. If any jolokia fails we will fail the scaledown. Thanks!

**Pull request merged**